### ISSUE TITLE
About GPU memory usage

### ISSUE THREAD

**Issue by Fan-Yixuan** [XuyangBai/TransFusion#10]
Thanks for your great work! I am trying to reimplement your work with the new version (v1.0.0) of mmd3d, my environment:

```
sys.platform: linux
Python: 3.8.11 (default, Aug  3 2021, 15:09:35) [GCC 7.5.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: GeForce RTX 3090
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.1, V11.1.74
GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 1.9.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.1.2 (Git Hash 98be7e8afa711dc9b66c8ff3504129cb82013cdb)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.0.5
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.9.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.10.1
OpenCV: 4.5.3
MMCV: 1.5.0
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.1
MMDetection: 2.23.0
MMSegmentation: 0.24.0
MMDetection3D: 1.0.0rc1+c7cde78
```

I have dealt with the coordinate system refactoring problem and also the `img_fields` issue, but I can only train with up to 50 query proposals while with one sample per 24GB RTX3090 GPU, using the default config (nuscenes, Lidar and camera, R50FPN, second lidar backbone, 200queries) will encounter CUDA OOM. 

Noting your practice https://github.com/XuyangBai/TransFusion/issues/6#issuecomment-1111812026, I hereby seek help. I also didn't notice your use of spconv, hope you can provide more details.
Thanks a lot.

**Comment by XuyangBai** [comment]
Hi @Fan-Yixuan Thanks for your interest in our work. I have tried training TransFusion on 8 3090GPUs and it could fit into the memory, so not sure what happens in your environment. But you could try to use spconv 1.2 to reduce the memory. The spconv is used in `SparseEncoder`. `mmdet3d` includes spconv in their repo in `mmdet3d/ops/spconv` but it is a old version. To use another version of spconv, I used to install it following the instruction [here](https://github.com/traveller59/spconv) and replace 
https://github.com/XuyangBai/TransFusion/blob/53370467c1b88f163cbe7b7300a1f588a6761e35/mmdet3d/ops/spconv/__init__.py#L14-L20

by something like 
```
from spconv import SparseConv2d, ...
```

**Comment by Fan-Yixuan** [comment]
@XuyangBai Hi dear author, I would like to ask if TransFusion's prediction heads do not contain branches for attribute prediction (moving, stopped, parked vehicle, etc.). I'm not familiar with this task (nuScenes), why does it work like this instead of reducing AAE by adding such branches.

**Comment by XuyangBai** [comment]
Ah sorry I just use it as the de-facto, never carefully think about this issue

**Comment by Fan-Yixuan** [comment]
Ok, since mmd3d implements it like this, it should have its own reason 2333333

**Comment by XuyangBai** [comment]
Did you use the newest code? There is some bugs when changing the shape of image features, leading to the mismatch between two modalities, which are fixed in https://github.com/XuyangBai/TransFusion/commit/8977b2b9ed74526b1ac4dd5ce9b60c195dc47056 and https://github.com/XuyangBai/TransFusion/commit/5187414d90d9a216f1e07520bf637a79f664132f

**Comment by XuyangBai** [comment]
`GlobalRotScaleTrans` and `RandomFlip3D` will not break the matching between LiDAR and camera because every time we project the object queries(and initial prediction) from 3d space onto the image plane, we first do the `inverse transformation`, which will convert the augmented 3d positions back to the original coordinate. See the following code: 
https://github.com/XuyangBai/TransFusion/blob/5187414d90d9a216f1e07520bf637a79f664132f/mmdet3d/models/dense_heads/transfusion_head.py#L944-L948

BTW, I just realize that this might be the problem: here I assume the BS=1 is for evaluation time so I skip the `apply_3d_transformation` for fast inference. If you use `samples_per_gpu=1` for training you should remove this logic and always apply `apply_3d_transformation`.

**Comment by XuyangBai** [comment]
Yes, It might be the reason. If the `flip` in `img_field` is set to True but the image is not flipped, then the consistency between Lidar and image is broken.

**Comment by Fan-Yixuan** [comment]
My concern is that maybe 
https://github.com/open-mmlab/mmdetection/blob/73b4e65a6a30435ef6a35f405e3474a4d9cfb234/mmdet/datasets/pipelines/transforms.py#L465-L469
should be changed to a loop over the list of images like you did in MyResize etc., but I don't understand why you and https://github.com/XuyangBai/TransFusion/issues/7#issuecomment-1118447999 are able to get the correct training results based on the current code.

**Comment by XuyangBai** [comment]
@Fan-Yixuan I find `mmcv.imflip` do works for list of images, see the following example:

![截屏2022-05-12 上午8 35 15](https://user-images.githubusercontent.com/15166089/167969178-af0bc902-1589-4784-abb8-e5cad55ffd3b.png)

**Comment by Fan-Yixuan** [comment]
Thanks for the explanation, that's true, but I still can't seem to solve my problem. The strangest thing I found is the change of loss_bbox during the training process as shown in the figure. The orange line is the result of LiDAR only, and the red line is the result of LC. Do you have any suggestions? thanks a lot
![2022-05-12 17-03-15 的屏幕截图](https://user-images.githubusercontent.com/45420358/168035497-ecee49a2-5078-4775-a426-5d27d057ba16.png)
.

**Comment by XuyangBai** [comment]
It is really weird that the bbox loss turns to increase at some point, the curve before 10k looks normal. I am not sure the reason but maybe you can first verify the projection of object queries onto the image through some visualization? If the lidar and image are not aligned well, the image feature attached to the object queries will be wrong. BTW, you mentioned the mATE, mASE, mAOE are all increasing, so how about mAP?

**Comment by Fan-Yixuan** [comment]
The first three epochs after adding camera, mAP: 62.49, 58.86, 59.66. I feel that the loss turns to increase is probably because the learning rate becomes larger (I use 4*3090 with 2 samples per GPU so I forward propagation twice and then update the parameters to make batch size equals 16, thus learning rate reaches a maximum at around 40k iters)

**Comment by XuyangBai** [comment]
The learning rate should not be the reason. I have also tried to use batch_size 8*1. 

Yes, I freeze the LiDAR branch during training of TransFusion as it is already well trained in the first stage. If you would like to jointly optimize the lidar branch and the fusion component, maybe they should be use different learning rate.

**Comment by Fan-Yixuan** [comment]
Hi, sorry for the late reply. I made two changes: the first follows your changes in the dataset definition file, but from what I understand this shouldn't have a real impact. 
![2022-05-14 10-34-48 的屏幕截图](https://user-images.githubusercontent.com/45420358/168407778-8d1b4112-1915-403b-9b5a-c6e51bce96b7.png)

The second is to freeze the weights of the LiDAR branch. Now I can get 66.75mAP/71.03NDS on the nuScenes validation set. So I think the previous problem is caused by using too large learning rate for the LiDAR branch which has been well trained.

**Comment by XuyangBai** [comment]
Yes, the order of images does not affect a lot but freezing the backbone did.

**Comment by Fan-Yixuan** [comment]
Ok, thank you for your patience and your excellent work, I close this issue.

**Comment by nmll** [comment]
@Fan-Yixuan  Hello! Could you tell me the max learning rate in your training step of  the first stage and second stage separately?

**Comment by nmll** [comment]
OK! Thanks!

**Comment by nmll** [comment]
> `GlobalRotScaleTrans` and `RandomFlip3D` will not break the matching between LiDAR and camera because every time we project the object queries(and initial prediction) from 3d space onto the image plane, we first do the `inverse transformation`, which will convert the augmented 3d positions back to the original coordinate. See the following code:
> 
> https://github.com/XuyangBai/TransFusion/blob/5187414d90d9a216f1e07520bf637a79f664132f/mmdet3d/models/dense_heads/transfusion_head.py#L944-L948
> 
> BTW, I just realize that this might be the problem: here I assume the BS=1 is for evaluation time so I skip the `apply_3d_transformation` for fast inference. If you use `samples_per_gpu=1` for training you should remove this logic and always apply `apply_3d_transformation`.


Hello! @XuyangBai  May I ask about this apply_3d_transformation is only used in projecting 3D to 2D query, but is not used in adding the BEV lidar feature and BEV image feature for image guided query initialization. Will this be a mismatch between lidar and image modalities due to the Radomflip3d and GlobalRotScaleTrans?

**Comment by heming7** [comment]
> 
Hello Yixuan

Can you tell what you did to solve the version issue? I am facing the same problem now.

**Comment by heming7** [comment]
> @XuyangBai Hi dear author, I finished training transfusion_nusc_voxel_L and got val set performance of 64.63mAP/69.99NDS. The previous problem about GPU memory has been solved, which was because the images were not resized to 448*800 due to some version issue. However, the training after adding cameras encountered some problems, mATE, mASE, mAOE, and mAVE are all increasing with training. Do you have any suggestion for possible causes? My question is, it seems that GlobalRotScaleTrans and RandomFlip3D will cause mismatch between LiDAR and camera?

Hello Yixuan

Can you tell what you did to solve the version issue? I am facing the same problem now.

**Comment by Fan-Yixuan** [comment]
Hi @heming7, you need to make sure that `results['img_fields'] is ['img']` and `type(results['img']) is list` before these code:
https://github.com/XuyangBai/TransFusion/blob/8977b2b9ed74526b1ac4dd5ce9b60c195dc47056/mmdet3d/datasets/pipelines/loading.py#L187-L190

**Comment by heminghuang7** [comment]
> 

Hello Yixuan

Thank you for the suggestion. I checked the code and I think the author has pushed a commit that fixes this. But I manage to run it by reducing the value samples_per_gpu. Anyway, thank you so much for the help!

**Comment by heminghuang7** [comment]
> Hi @heming7, you need to make sure that `results['img_fields'] is ['img']` and `type(results['img']) is list` before these code:
> 
> https://github.com/XuyangBai/TransFusion/blob/8977b2b9ed74526b1ac4dd5ce9b60c195dc47056/mmdet3d/datasets/pipelines/loading.py#L187-L190

Hello Yixuan

Thank you for the suggestion. I checked the code and I think the author has pushed a commit that fixes this. But I manage to run it by reducing the value samples_per_gpu. Anyway, thank you so much for the help!

**Comment by yinjunbo** [comment]
> @XuyangBai Hi dear author, I finished training transfusion_nusc_voxel_L and got val set performance of 64.63mAP/69.99NDS. The previous problem about GPU memory has been solved, which was because the images were not resized to 448*800 due to some version issue. However, the training after adding cameras encountered some problems, mATE, mASE, mAOE, and mAVE are all increasing with training. Do you have any suggestion for possible causes? My question is, it seems that GlobalRotScaleTrans and RandomFlip3D will cause mismatch between LiDAR and camera?

Hi, @Fan-Yixuan, could you please share your torch/cuda/mmdet3d/spconv environment you've used  to reproduce the nusc val performance (64.63mAP and 69.99NDS)? It seems that you used 8*3090 with batch size 2 and lr 1e-4?

**Comment by Fan-Yixuan** [comment]
> > @XuyangBai Hi dear author, I finished training transfusion_nusc_voxel_L and got val set performance of 64.63mAP/69.99NDS. The previous problem about GPU memory has been solved, which was because the images were not resized to 448*800 due to some version issue. However, the training after adding cameras encountered some problems, mATE, mASE, mAOE, and mAVE are all increasing with training. Do you have any suggestion for possible causes? My question is, it seems that GlobalRotScaleTrans and RandomFlip3D will cause mismatch between LiDAR and camera?
> 
> Hi, @Fan-Yixuan, could you please share your torch/cuda/mmdet3d/spconv environment you've used to reproduce the nusc val performance (64.63mAP and 69.99NDS)? It seems that you used 8*3090 with batch size 2 and lr 1e-4?

Hi my env: https://github.com/XuyangBai/TransFusion/issues/10#issue-1226416264, my spconv: 2.1.21 my total batchsize: 16, lr: 1e-4

**Comment by yinjunbo** [comment]
> > > @XuyangBai Hi dear author, I finished training transfusion_nusc_voxel_L and got val set performance of 64.63mAP/69.99NDS. The previous problem about GPU memory has been solved, which was because the images were not resized to 448*800 due to some version issue. However, the training after adding cameras encountered some problems, mATE, mASE, mAOE, and mAVE are all increasing with training. Do you have any suggestion for possible causes? My question is, it seems that GlobalRotScaleTrans and RandomFlip3D will cause mismatch between LiDAR and camera?
> > 
> > 
> > Hi, @Fan-Yixuan, could you please share your torch/cuda/mmdet3d/spconv environment you've used to reproduce the nusc val performance (64.63mAP and 69.99NDS)? It seems that you used 8*3090 with batch size 2 and lr 1e-4?
> 
> Hi my env: [#10 (comment)](https://github.com/XuyangBai/TransFusion/issues/10#issue-1226416264), my spconv: 2.1.21 my total batchsize: 16, lr: 1e-4
@Fan-Yixuan , Thanks for your quik reply! I'll have another try.
Btw, could you please share your traing log, so I can check my problem accordingly(email: yinjunbocn@gmail.com)?

**Comment by Fan-Yixuan** [comment]
@yinjunbo 
Sure, for training lidar-only, the first 15 epochs:
[20220505_225100.log](https://github.com/XuyangBai/TransFusion/files/9743831/20220505_225100.log)
the last 5 epochs (fade strategy):
[20220508_101828.log](https://github.com/XuyangBai/TransFusion/files/9743844/20220508_101828.log)

**Comment by yinjunbo** [comment]
> @yinjunbo Sure, for training lidar-only, the first 15 epochs: [20220505_225100.log](https://github.com/XuyangBai/TransFusion/files/9743831/20220505_225100.log) the last 5 epochs (fade strategy): [20220508_101828.log](https://github.com/XuyangBai/TransFusion/files/9743844/20220508_101828.log)

Thank you very much!
I find that my training loss is obvisously larger than yours. Did you try to train a model before coordinate system refactoring ?

**Comment by Fan-Yixuan** [comment]
@yinjunbo Sorry I didn't save the training logs before modifying the coordinate system, but if the coordinate is not aligned, it should work very poorly.

**Comment by yinjunbo** [comment]
> @yinjunbo Sorry I didn't save the training logs before modifying the coordinate system, but if the coordinate is not aligned, it should work very poorly.

I totally agree with you. Since my repreoced performance is just slightly lower (~2 points) than yours, this could not be caused by coordinate system. I'll continue to find the problems. tks!

**Comment by BoomSky0416** [comment]
@Fan-Yixuan Hello, I am trying to reproduce transfusion in mmdet3d-1.1.0. But I got the wrong result in training lidar-camera fusion stage. Could you please share your training log for this stage, thanks! (email: shoutian@umich.edu)

**Comment by Arthav24** [comment]
Hi, @Fan-Yixuan, I’m also trying to reimplement the original author’s work using the updated mmdetection3d (1.0.0rc6). I managed to run the L training, but I’m encountering a few issues while training LC, especially with the ‘scale_factor’ parameter. Could you please share your changes or the repository? I’m also interested in verifying the coordinate system refactoring. Thanks!
